mobile-sdk/src/main/java/com/github/itwin/mobilesdk/jsonvalue/JSONValue.kt

```diff
@@ -0,0 +1,406 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+@file:Suppress("unused")
+package com.github.itwin.mobilesdk.jsonvalue
+
+import org.json.JSONArray
+import org.json.JSONObject
+
+/**
+ * Class to represent a top-level JSON object. This can be an object, array, string, number,
+ * boolean, or null.
+ * @param value The value from which to create a [JSONValue]. Must be `null` or a supported type:
+ * [Boolean], [String], [Number], [JSONArray], or [JSONObject]. If you need a [JSONValue] built from
+ * any other type, use [toJSON].
+ */
+class JSONValue(value: Any? = null) {
```

Comment:
I suggest JSONValue can be more type-safe by only allowing construction with specific values.

```suggestion
class JSONValue private constructor(value: Any? = null) {
    constructor() : this(null as Any?)
    constructor(value: JSONObject) : this(value as Any)
    constructor(value: JSONArray) : this(value as Any)
    constructor(value: String) : this(value as Any)
    constructor(value: Number) : this(value as Any)
    constructor(value: Boolean) : this(value as Any)

```

Reply:
Fixed. Note that this can change how the class must be used (see changes in toJSON), due to the individual constructors not being equivalent. Note also that all of the type-specific constructors must be optional, otherwise optional values must be checked for null and then have the 0-argument constructor used if they are null.